src/backend/distributed/master/master_metadata_utility.c

```diff
@@ -57,14 +61,274 @@ static void RecordDistributedRelationDependencies(Oid distributedRelationId,
+		WorkerNode *workerNode = (WorkerNode *) lfirst(workerNodeCell);
+		uint64 relationSizeOnNode = DistributedTableSizeOnWorker(workerNode, relationId,
+																 sizeQuery);
+		totalRelationSize += relationSizeOnNode;
+	}
+
+	heap_close(pgDistNode, NoLock);
+	return totalRelationSize;
+}
+
+
+/*
+ * DistributedTableSizeOnWorker gets the workerNode and relationId to calculate size
+ * of that relation on the given workerNode by summing up the size of each shard placement.
+ */
+static uint64
+DistributedTableSizeOnWorker(WorkerNode *workerNode, Oid relationId, char *sizeQuery)
+{
+	StringInfo tableSizeQuery = NULL;
+	char *workerNodeName = workerNode->workerName;
+	uint32 workerNodePort = workerNode->workerPort;
+	StringInfo tableSizeStringInfo;
+	char *tableSizeString;
+	List *sizeList = NIL;
+	uint64 relationSize = 0;
+	List *shardIntervalsOnNode = ShardIntervalsOnWorkerNode(workerNode, relationId);
+
+	tableSizeQuery = GenerateSizeQueryOnMultiplePlacements(relationId,
+														   shardIntervalsOnNode,
+														   sizeQuery);
+
+	sizeList = ExecuteRemoteQuery(workerNodeName, workerNodePort, NULL, tableSizeQuery);
+
+	if (sizeList == NIL)
+	{
+		ereport(ERROR, (errcode(ERRCODE_CONNECTION_FAILURE),
+						errmsg("cannot get the size")));
+	}
+
+	tableSizeStringInfo = (StringInfo) linitial(sizeList);
+	tableSizeString = tableSizeStringInfo->data;
+	relationSize = atol(tableSizeString);
+
+	return relationSize;
+}
+
+
+/*
+ * ShardIntervalsOnNode takes a WorkerNode then compares it with each placement of table.
+ * It returns shard intervals of table on that node as a list of shard intervals.
+ * Note that, shard intervals returned as elements of the list are not the copies but the
+ * pointers. DO NOT modify the shard intervals returned by this function.
+ */
+static List *
+ShardIntervalsOnWorkerNode(WorkerNode *workerNode, Oid relationId)
+{
+	DistTableCacheEntry *distTableCacheEntry = DistributedTableCacheEntry(relationId);
+	char *workerNodeName = workerNode->workerName;
+	uint32 workerNodePort = workerNode->workerPort;
+	List *shardIntervalList = NIL;
+	int shardIndex = 0;
+	int shardIntervalArrayLength = distTableCacheEntry->shardIntervalArrayLength;
+
+	for (shardIndex = 0; shardIndex < shardIntervalArrayLength; shardIndex++)
+	{
+		ShardPlacement *placementArray =
+			distTableCacheEntry->arrayOfPlacementArrays[shardIndex];
+		int numberOfPlacements =
+			distTableCacheEntry->arrayOfPlacementArrayLengths[shardIndex];
+		int placementIndex = 0;
+
+		for (placementIndex = 0; placementIndex < numberOfPlacements; placementIndex++)
+		{
+			ShardPlacement *placement = &placementArray[placementIndex];
+			char *shardNodeName = placement->nodeName;
+			uint32 shardNodePort = placement->nodePort;
+			uint64 shardId = placement->shardId;
+			bool metadataLock = false;
+
+			metadataLock = TryLockShardDistributionMetadata(shardId, AccessShareLock);
+
+			/* if the lock is not acquired warn the user */
+			if (metadataLock == false)
```

Comment:
`AccessShareLock` should be `ShareLock`. `AccessShareLock` has no effect since we never obtain an `AccessExclusiveLock`.
https://www.postgresql.org/docs/current/static/explicit-locking.html 

This change made me realise the metadata locking is kind of broken at the moment (though this is fine, since this function is best-effort). I opened #1280.